Add Vision to the Robot

a2aaron opened the pull request.

a2aaron committed:
Merged branch master into transfer/vision

a2aaron committed:
Add Vision classes to the Robot

a2aaron committed:
Add Vision Debugging to the SmartDashboard

porglezomp commented on src/org/usfirst/frc/team1076/robot/Robot.java:
```diff
@@ -43,6 +45,13 @@ public void robotInit() {
 		chooser.addDefault("Default Auto", new ExampleCommand());
 		// chooser.addObject("My Auto", new MyAutoCommand());
 		SmartDashboard.putData("Auto mode", chooser);
+        SmartDashboard.putNumber("Show Vision", 1);
```
We should use `SmarterDashboard` for this!

a2aaron replied:
We can do this when we merge the SmarterDashboard commits from 2017-Pickup.

porglezomp commented on src/org/usfirst/frc/team1076/robot/Robot.java:
```diff
@@ -53,11 +62,18 @@ public void robotInit() {
 		Scheduler.getInstance().run();
+		if (debugCount++ % 100 == 0 && SmartDashboard.getNumber("Show Vision", 0) == 1) {
+		    receiver.receive();
+		    if (receiver == null) {
```
If receiver was `null`, we would get a NPE on the previous line.

porglezomp commented on src/org/usfirst/frc/team1076/robot/vision/VisionData.java:
```diff
@@ -0,0 +1,116 @@
+            heading = json.getInt("heading");
+            range = json.getDouble("range");
+        } catch (JSONException e) { 
+            // Malformed JSON packet, assume no useful data
```
When we get to this point, we might have already set some data and we should probably clear it.

porglezomp commented on src/org/usfirst/frc/team1076/robot/vision/VisionData.java:
```diff
@@ -0,0 +1,116 @@
+    public void update(String json) {
+    	try {
+    		update(new JSONObject(new JSONTokener(json)));
+    	} catch (JSONException e) {
```
We should perhaps set the status to error?

porglezomp reviewed (changes requested):
Definitely fix the NPE bug, the others are probably helpful for robustness too.

a2aaron committed:
Fix potential NPE in the vision debug messages.

If receiver is null, then trying to do receiver.receive() will cause an NPE.

a2aaron committed:
Prevent VisionData values from being updated from a bad JSON packet.

If a JSON packet happens to be cutoff part way through (Say, it doesn't have a range field), then `update()` will update some but not all of the fields, which is undesireable. Instead, we check that all fields exist before assigning them.

a2aaron committed:
Ensure that all `update()` functions set status to ERROR for bad JSON

a2aaron committed:
Merge branch 'master' into transfer/vision

a2aaron committed:
Merge branch 'master' into transfer/vision

porglezomp commented on src/org/usfirst/frc/team1076/robot/Robot.java:
```diff
@@ -46,14 +47,13 @@ public void robotInit() {
-		SmartDashboard.putData("Auto mode", chooser);
-        SmartDashboard.putNumber("Show Vision", 1);
-
+    SmarterDashboard.putNumber("Show Vision", 1);
```
This needs to be `putDefaultNumber`

porglezomp reviewed (changes requested):
Missed the `putDefaultNumber`

a2aaron committed:
Use SmarterDashboard and not SmartDashboard
